### PR TITLE
Support gs default application credentials.

### PR DESCRIPTION
**Pull request by andrewchambers**
address #28

Up for review, It is *not* suitable for merge until the TODO is addressed.

### CODE REVIEW

File: src/s3ql/backends/gs.py [hunk 1]
```diff
@@ -53,6 +53,14 @@ def __init__(self, options):
         if self.use_oauth2:
             self.hdr_prefix = 'x-goog-'
 
+        self.google_app_credentials = None
+        if self.password == '':
+            log.info('Password empty, using google application default account')
```
Comment (Nikratio) [inline src/s3ql/backends/gs.py]: The objects will be instantiated more than once, it's ugly to have it log this every time. Please either change to `log.debug`, or add a `LOG_ONCE` stanza.

File: src/s3ql/backends/gs.py [hunk 2]
```diff
@@ -53,6 +53,14 @@ def __init__(self, options):
...
+            # Don't import at top level for backwards compatibility
+            import google.auth
```
Comment (Nikratio) [inline src/s3ql/backends/gs.py]: I'd rather have a condition import at the top and only use the module if it's available. Having it crash with an ImportError is ugly.

File: src/s3ql/backends/gs.py [hunk 3]
```diff
@@ -172,6 +184,25 @@ def _get_access_token(self):
         finally:
             conn.disconnect()
 
+    def _refresh_application_default_token(self):
+        # For backwards compatibility we only use this token method if the
+        # password is empty.
+        assert(self.password == '')
```
Comment (Nikratio) [inline src/s3ql/backends/gs.py]: No parens, assert is a statement.

**Comment by Nikratio** [comment]
Looks ok so far, but the TODO is a big if. If that requires a huge chunk of code with multiple extra imports, then I'm not that enthusiastic...

**Comment by andrewchambers** [comment]
Thanks for the feedback. I'll address the issues and give fixing the TODO a try to see how big it is. While I would like it mainlined, I'm using this code on an internal project that runs on google cloud, so its not that bad for me to just keep it as an internal patch set.

**Comment by Nikratio** [comment]
(will merge alternate implementation)

**Pull request closed without merging**